eo-phi-normalizer/src/Language/EO/Phi/Rules/Yaml.hs

```diff
@@ -0,0 +1,207 @@
+
+  Rules.GlobalDispatch a -> do
+    a' <- applySubstAttr subst a
+    return (Phi.GlobalDispatch a')
+
+  Rules.ThisDispatch a -> do
+    a' <- applySubstAttr subst a
+    return (Phi.ThisDispatch a')
+
+  Rules.Termination -> return Phi.Termination
+
+  Rules.MetaObject x -> lookup x objectMetas
+
+applySubstAttr :: Subst -> Rules.Attribute -> Maybe (Phi.Attribute)
+applySubstAttr Subst{..} = \case
+  Rules.Phi     -> return Phi.Phi
+  Rules.Rho     -> return Phi.Rho
+  Rules.Sigma   -> return Phi.Sigma
+  Rules.VTX     -> return Phi.VTX
+  Rules.Label l -> return (Phi.Label (coerce l))
+  Rules.Alpha a -> return (Phi.Alpha (coerce a))
+  Rules.MetaAttr x -> lookup x attributeMetas
+
+applySubstBindings :: Subst -> [Rules.Binding] -> Maybe [Phi.Binding]
+applySubstBindings subst bindings =
+  concat <$> mapM (applySubstBinding subst) bindings
+
+applySubstBinding :: Subst -> Rules.Binding -> Maybe [Phi.Binding]
+applySubstBinding subst@Subst{..} = \case
+  Rules.AlphaBinding a obj -> do
+    a' <- applySubstAttr subst a
+    obj' <- applySubst subst obj
+    return [Phi.AlphaBinding a' obj']
+  Rules.EmptyBinding a -> do
+    a' <- applySubstAttr subst a
+    return [Phi.EmptyBinding a']
+  Rules.DeltaBinding bytes -> return [Phi.DeltaBinding (coerce bytes)]
+  Rules.LambdaBinding bytes -> return [Phi.LambdaBinding (coerce bytes)]
+  Rules.MetaBindings m -> lookup m bindingsMetas
+
+mergeSubst :: Subst -> Subst -> Subst
+mergeSubst (Subst xs ys zs) (Subst xs' ys' zs') =
+  Subst (xs ++ xs') (ys ++ ys') (zs ++ zs')
+
+-- 1. need to implement applySubst' :: Subst -> Rules.Object -> Rules.Object
+-- 2. complete the code
+matchObject :: Rules.Object -> Phi.Object -> [Subst]
+matchObject (Rules.Formation ps) (Phi.Formation bs) =
+  matchBindings ps bs
+matchObject (Rules.Application obj bindings) (Phi.Application obj' bindings') = do
+  subst1 <- matchObject obj obj'
+  subst2 <- matchBindings (applySubstBindings' subst1 bindings) bindings'
+  return (subst1 <> subst2)
+matchObject (Rules.ObjectDispatch pat a) (Phi.ObjectDispatch obj a') = do
+  subst1 <- matchObject pat obj
+  subst2 <- matchAttr (applySubstAttr' subst1 a) a'
+  return (subst1 <> subst2)
+matchObject (Rules.MetaObject m) obj = return Subst
+  { objectMetas = [(m, obj)], bindingsMetas = [], attributeMetas = [] }
+
+matchBindings :: [Rules.Binding] -> [Phi.Binding] -> [Subst]
+matchBindings [] [] = [emptySubst]
+matchBindings [Rules.MetaBindings b] bindings = return Subst
+  { objectMetas = [], bindingsMetas = [(b, bindings)], attributeMetas = [] }
+matchBindings (p:ps) bs = do
+  (bs', subst1) <- matchFindBinding p bs
+  subst2 <- matchBindings ps bs'
+  return (subst1 <> subst2)
+
+-- >>> select [1,2,3,4]
+-- [(1,[2,3,4]),(2,[1,3,4]),(3,[1,2,4]),(4,[1,2,3])]
+select :: [a] -> [(a, [a])]
+select [] = []
+select [x] = [(x, [])]
+select (x:xs) = (x, xs) :
+  [ (y, x:ys)
+  | (y, ys) <- select xs
+  ]
+
+matchFindBinding :: Rules.Binding -> [Phi.Binding] -> [([Phi.Binding], Subst)]
```

Comment:
I think we shouldn't be non-deterministic here as it can badly affect the performance (IDK if laziness will save us).

So, maybe apply `listToMaybe` instead of returning a list?

Reply:
Sensible suggestion **if** we expect at most one match. However, this is not currently the case if `Rule.Binding` contains a metavariable binding (i.e. we should match any metavariable). To make this efficient we should then somehow stop and let other part of the matching figure out what metavariable we are talking about, which can be done in some places by rearranging order of matching (for example, for `ObjectDispatch`). However, this is not possible in general without constraint-based matching.